specreduce/wavelength_calibration.py

```diff
@@ -0,0 +1,228 @@
+            Can also be input as an `~astropy.table.QTable` table with (minimally) a column
+            named "pixel_center" and optionally a "wavelength" column with known line
+            wavelengths populated.
+        line_wavelengths: `~astropy.units.Quantity`, `~astropy.table.QTable`, optional
+            `astropy.units.Quantity` array of line wavelength values corresponding to the
+            line_pixels. Does not have to be in the same order (the lists will be sorted)
+            but does currently need to be the same length as line_list. Can also be input
+            as an `~astropy.table.QTable` with (minimally) a "wavelength" column.
+        catalog: list, str, optional
+            The name of a catalog of line wavelengths to load and use in automated and
+            template-matching line matching.
+        model: `~astropy.modeling.Model`
+            The model to fit for the wavelength solution. Defaults to a linear model.
+        """
+        self._input_spectrum = input_spectrum
+        self._model = model
+        self._line_list = line_list
+        self.spectral_unit = spectral_unit
+        self._cached_properties = ['wcs',]
+        self.fitter = fitter
+        self._potential_wavelengths = None
+        self._catalog = catalog
+
+        # ToDo: Implement having line catalogs
+        self._available_catalogs = get_available_catalogs()
+
+        if isinstance(line_list, (list, np.ndarray)):
+            self._line_list = QTable([line_list], names=["pixel_center"])
+
+        if self._line_list["pixel_center"].unit is None:
+            self._line_list["pixel_center"].unit = u.pix
+
+        # Make sure our pixel locations are sorted
+        self._line_list.sort("pixel_center")
+
+        if (line_wavelengths is None and catalog is None
+                and "wavelength" not in self._line_list.columns):
+            raise ValueError("You must specify at least one of line_wavelengths, "
+                             "catalog, or 'wavelength' column in line_list.")
+
+        # Sanity checks on line_wavelengths value
+        if line_wavelengths is not None:
+            if "wavelength" in line_list:
+                raise ValueError("Cannot specify line_wavelengths separately if there is"
+                                 "a 'wavelength' column in line_list.")
+            if len(line_wavelengths) != len(line_list):
+                raise ValueError("If line_wavelengths is specified, it must have the same "
+                                 "length as line_pixels")
+            if not isinstance(line_wavelengths, (u.Quantity, QTable)):
+                raise ValueError("line_wavelengths must be specified as an astropy.units.Quantity"
+                                 "array or as an astropy.table.QTable")
+            if isinstance(line_wavelengths, u.Quantity):
+                line_wavelengths.value.sort()
+                self._line_list["wavelength"] = line_wavelengths
```

Comment:
does `line_wavelengths` support frequencies?  Would the sorting need to be flipped in this case?  What if they're mixed units?

Reply:
I'm not sure how common it is to use frequencies for this (my assumption was "never" but I guess _someone_ might want to), but we could support it. I can also add a check to make sure they're all in the same unit.

Reply:
Actually, since this requires `line_wavelengths` to be a `Quantity` or `QTable` (rather than allowing a list of `Quantity`), it's already guaranteed that all elements have the same unit.

Reply:
@kecnry I implemented reverse-sorting if frequencies are provided, although that does make the variable name `line_wavelengths` a bit of a misnomer.

Reply:
my first take is that since this package is geared towards 2D optical/IR data obtained with CCD-like devices, i would stick with wavelengths internally and not worry about frequencies at all. at least for simplicity at this stage of development. if users have dispersed data with units of frequency, there are existing utilities to transform it to wavelengths. i can imagine possible use cases for 2D spatial vs frequency data in radio astronomy, but i've never seen it. radio data is either 1D (single-dish) or 3D (scanned single-dish or interferometry). eV or wavenumber might be more likely candidates for non-wavelength inputs, but they can also easily be transformed to wavelength.